src/cargo/core/profiles.rs

```diff
@@ -0,0 +1,394 @@
+            incremental: false,
+            panic: None,
+        }
+    }
+}
+
+impl fmt::Display for Profile {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Profile({})", self.name)
+    }
+}
+
+impl hash::Hash for Profile {
+    fn hash<H>(&self, state: &mut H)
+    where
+        H: hash::Hasher,
+    {
+        self.comparable().hash(state);
+    }
+}
+
+impl cmp::PartialEq for Profile {
+    fn eq(&self, other: &Self) -> bool {
+        self.comparable() == other.comparable()
+    }
+}
+
+impl Profile {
+    fn default_dev() -> Profile {
+        Profile {
+            name: "dev",
+            debuginfo: Some(2),
+            debug_assertions: true,
+            overflow_checks: true,
+            incremental: true,
+            ..Profile::default()
+        }
+    }
+
+    fn default_release() -> Profile {
+        Profile {
+            name: "release",
+            opt_level: InternedString::new("3"),
+            ..Profile::default()
+        }
+    }
+
+    fn default_test() -> Profile {
+        Profile {
+            name: "test",
+            ..Profile::default_dev()
+        }
+    }
+
+    fn default_bench() -> Profile {
+        Profile {
+            name: "bench",
+            ..Profile::default_release()
+        }
+    }
+
+    fn default_doc() -> Profile {
+        Profile {
+            name: "doc",
+            ..Profile::default_dev()
+        }
+    }
+
+    /// Compare all fields except `name`, which doesn't affect compilation.
+    /// This is necessary for `Unit` deduplication for things like "test" and
+    /// "dev" which are essentially the same.
+    fn comparable(
+        &self,
+    ) -> (
+        &InternedString,
+        &Lto,
+        &Option<u32>,
+        &Option<u32>,
+        &bool,
+        &bool,
+        &bool,
+        &bool,
+        &Option<InternedString>,
+    ) {
+        (
+            &self.opt_level,
+            &self.lto,
+            &self.codegen_units,
+            &self.debuginfo,
+            &self.debug_assertions,
+            &self.overflow_checks,
+            &self.rpath,
+            &self.incremental,
+            &self.panic,
+        )
+    }
+}
+
+/// The link-time-optimization setting.
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+pub enum Lto {
+    /// False = no LTO
+    /// True = "Fat" LTO
+    Bool(bool),
+    /// Named LTO settings like "thin".
+    Named(String),
+}
+
+/// A flag used in `Unit` to indicate the purpose for the target.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum ProfileFor {
+    /// A general-purpose target.
+    Any,
+    /// A target for `build.rs` or any of its dependencies.  This enables
+    /// `build_override` profiles for these targets.
+    CustomBuild,
+    /// A target that is a dependency of a test or benchmark.  Currently this
+    /// enforces that the `panic` setting is not set.
+    TestDependency,
+}
+
+impl ProfileFor {
+    pub fn all_values() -> Vec<ProfileFor> {
```

Comment:
Mincooptimization, but here and in `CompileMode::all_modes`, we could return `&'static [Stuff]`, to save some allocations.